File: app/src/main/java/com/softbankrobotics/qisdktutorials/ui/tutorials/peoplecharacteristics/EmotionTutorialActivity.java
package com.softbankrobotics.qisdktutorials.ui.tutorials.peoplecharacteristics;

import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.widget.ImageView;

import com.aldebaran.qi.sdk.QiContext;
import com.aldebaran.qi.sdk.QiSDK;
import com.aldebaran.qi.sdk.RobotLifecycleCallbacks;
import com.aldebaran.qi.sdk.builder.SayBuilder;
import com.aldebaran.qi.sdk.object.conversation.Say;
import com.softbankrobotics.qisdktutorials.R;
import com.softbankrobotics.qisdktutorials.model.data.BasicEmotion;
import com.softbankrobotics.qisdktutorials.model.observer.BasicEmotionObserver;
import com.softbankrobotics.qisdktutorials.ui.conversation.ConversationItemType;
import com.softbankrobotics.qisdktutorials.ui.conversation.ConversationView;
import com.softbankrobotics.qisdktutorials.ui.tutorials.TutorialActivity;

/**
 * The activity for the Emotion tutorial.
 */
public class EmotionTutorialActivity extends TutorialActivity implements RobotLifecycleCallbacks, OnBasicEmotionChangedListener {

    // Store the basic emotion observer.
    private BasicEmotionObserver basicEmotionObserver;

    private ConversationView conversationView;
    private ImageView emotionView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        conversationView = findViewById(R.id.conversationView);
        emotionView = findViewById(R.id.emotionView);

        // Create the basic emotion observer and listen to it.
        basicEmotionObserver = new BasicEmotionObserver();
        basicEmotionObserver.setListener(this);

        // Register the RobotLifecycleCallbacks to this Activity.
        QiSDK.register(this, this);
    }

    @Override
    protected void onDestroy() {
        // Stop listening to basic emotion observer and remove it.
        basicEmotionObserver.setListener(null);
        basicEmotionObserver = null;

        // Unregister all the RobotLifecycleCallbacks for this Activity.
        QiSDK.unregister(this);
        super.onDestroy();
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_emotion_tutorial;
    }

    @Override
    public void onRobotFocusGained(QiContext qiContext) {
        String textToSay = "I can display the basic emotions of the human I'm seeing. Try to express an emotion with your smile, your voice or by touching my sensors.";
        displayLine(textToSay, ConversationItemType.ROBOT_OUTPUT);

        Say say = SayBuilder.with(qiContext)
                .withText(textToSay)
                .build();

        say.run();

        // Start the basic emotion observation.
        basicEmotionObserver.startObserving(qiContext);
    }

    @Override
    public void onRobotFocusLost() {
        // Stop the basic emotion observation.
        basicEmotionObserver.stopObserving();
    }

    @Override
    public void onRobotFocusRefused(String reason) {
        // Nothing here.
    }

    @Override
    public void onBasicEmotionChanged(final BasicEmotion basicEmotion) {
        // Update basic emotion image.
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                emotionView.setImageResource(emotionImageRes(basicEmotion));
            }
        });
    }

    @DrawableRes
    private int emotionImageRes(BasicEmotion basicEmotion) {
        switch (basicEmotion) {
            case UNKNOWN:
                return R.drawable.ic_icons_cute_anon_unknown;
            case NEUTRAL:
                return R.drawable.ic_icons_cute_anon_neutral;
            case CONTENT:
                return R.drawable.ic_icons_cute_anon_smile;
            case JOYFUL:
                return R.drawable.ic_icons_cute_anon_joyful;
            case SAD:
                return R.drawable.ic_icons_cute_anon_sad;
            case ANGRY:
                return R.drawable.ic_icons_cute_anon_anger;
            default:
                throw new IllegalArgumentException("Unknown basic emotion: " + basicEmotion);
        }
    }

    private void displayLine(final String text, final ConversationItemType type) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                conversationView.addLine(text, type);
            }
        });
    }
}

File: app/src/main/java/com/softbankrobotics/qisdktutorials/ui/tutorials/perceptions/emotiondetection/EmotionTutorialActivity.kt
/*
 * Copyright (C) 2018 Softbank Robotics Europe
 * See COPYING for the license
 */

package com.softbankrobotics.qisdktutorials.ui.tutorials.perceptions.emotiondetection

import android.os.Bundle
import androidx.annotation.DrawableRes

import com.aldebaran.qi.sdk.QiContext
import com.aldebaran.qi.sdk.QiSDK
import com.aldebaran.qi.sdk.RobotLifecycleCallbacks
import com.aldebaran.qi.sdk.builder.SayBuilder
import com.softbankrobotics.qisdktutorials.R
import com.softbankrobotics.qisdktutorials.ui.conversation.ConversationBinder
import com.softbankrobotics.qisdktutorials.ui.tutorials.TutorialActivity
import kotlinx.android.synthetic.main.activity_emotion_tutorial.*

/**
 * The activity for the Emotion tutorial.
 */
class EmotionTutorialActivity : TutorialActivity(), RobotLifecycleCallbacks, OnBasicEmotionChangedListener {

    // Store the basic emotion observer.
    private var basicEmotionObserver: BasicEmotionObserver? = null

    private lateinit var conversationBinder: ConversationBinder

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Create the basic emotion observer and listen to it.
        basicEmotionObserver = BasicEmotionObserver()
        basicEmotionObserver?.listener = this

        // Register the RobotLifecycleCallbacks to this Activity.
        QiSDK.register(this, this)
    }

    override fun onDestroy() {
        // Stop listening to basic emotion observer and remove it.
        basicEmotionObserver?.listener = null
        basicEmotionObserver = null

        // Unregister the RobotLifecycleCallbacks for this Activity.
        QiSDK.unregister(this, this)
        super.onDestroy()
    }

    override val layoutId = R.layout.activity_emotion_tutorial

    override fun onRobotFocusGained(qiContext: QiContext) {
        // Bind the conversational events to the view.
        val conversationStatus = qiContext.conversation.status(qiContext.robotContext)
        conversationBinder = conversation_view.bindConversationTo(conversationStatus)

        val say = SayBuilder.with(qiContext)
                .withText("I can display the basic emotions of the human I'm seeing. Try to express an emotion with your smile, your voice or by touching my sensors.")
                .build()

        say.run()

        // Start the basic emotion observation.
        basicEmotionObserver?.startObserving(qiContext)
    }

    override fun onRobotFocusLost() {
        conversationBinder.unbind()

        // Stop the basic emotion observation.
        basicEmotionObserver?.stopObserving()
    }

    override fun onRobotFocusRefused(reason: String) {
        // Nothing here.
    }

    override fun onBasicEmotionChanged(basicEmotion: BasicEmotion) {
        // Update basic emotion image.
        runOnUiThread { emotion_image_view.setImageResource(emotionImageRes(basicEmotion)) }
    }

    @DrawableRes
    private fun emotionImageRes(basicEmotion: BasicEmotion): Int {
        return when (basicEmotion) {
            BasicEmotion.UNKNOWN -> R.drawable.ic_icons_cute_anon_unknown
            BasicEmotion.NEUTRAL -> R.drawable.ic_icons_cute_anon_neutral
            BasicEmotion.CONTENT -> R.drawable.ic_icons_cute_anon_smile
            BasicEmotion.JOYFUL -> R.drawable.ic_icons_cute_anon_joyful
            BasicEmotion.SAD -> R.drawable.ic_icons_cute_anon_sad
            BasicEmotion.ANGRY -> R.drawable.ic_icons_cute_anon_anger
        }
    }
}
